Add NotAuthorized component to handle 403s

akofink opened the pull request:
Stolen (and modified a bit) directly from https://github.com/RedHatInsights/insights-rbac-ui/blob/master/src/presentational-components/states/DeniedState.js

Signed-off-by: Andrew Kofink <akofink@redhat.com>

dLobatog commented on packages/components/src/Components/NotAuthorized/NotAuthorized.js:
```diff
@@ -0,0 +1,39 @@
+    Title,
+    Button
+} from '@patternfly/react-core';
+import { withRouter } from 'react-router-dom';
```
I think that you could do this without props with `useHistory` (https://github.com/ReactTraining/react-router/blob/master/packages/react-router/docs/api/hooks.md#usehistory) - you'd need to update react-router-dom to 5.x, I don't know if that's OK though.

dLobatog replied:
Is react-router-dom needed at all here actually? You can call `history.back()` directly from the browser without importing anything

bastilian commented:
@akofink Looks good to me! When linking (not npm linking) `react-router-dom` in compliance to the package under node_modules in frontend-components the `NotAuthorized` component renders normally in the compliance app and does not warn about `You should not use <Route> or withRouter() outside a <Router>`. It should also render normaly when the `frontend-components` package is release with this component and normally installed via `npm`.

I'm not sure why that is. Maybe someone else in @RedHatInsights/ui-reviewers has seen this issue before and can shed some light on this.

bastilian commented:
Ps.: I did try to upgrade `react-router-dom` packages to the latest and the same version in all related packages, but that didn't do the trick. It had to be exacatly the same package. aparently. 🤷

karelhala commented on packages/components/src/Components/NotAuthorized/Icon403.js:
```diff
@@ -0,0 +1,10 @@
+import React from 'react';
```
Please don't use this component, it is already exported by PF for us
```JSX
import { LockIcon } from '@patternfly/react-icons`;
```

you can check the full list of icons at https://patternfly-react.surge.sh/icons/

karelhala commented on packages/components/src/Components/NotAuthorized/NotAuthorized.js:
```diff
@@ -0,0 +1,42 @@
+            Return to previous page
+        </Button>;
+    } else {
+        button = <Button onClick={ history.push('/') }>
```
This will operate only at app level. Meaning if I go from `insights/compliance/systems` to `insights/compliance/systems/UUID` I can go back to `insights/compliance/systems` if you'll use `history.push('/')` it'll redirect you to `insights/compliance/`.

We've had a consensus with UX folks to redirect users to root URL `window.location.href = './'`

karelhala commented on packages/components/src/Components/NotAuthorized/NotAuthorized.js:
```diff
@@ -0,0 +1,42 @@
+import { useHistory } from 'react-router-dom';
+
+const NotAuthorized = ({ serviceName, ...props }) => {
+    let history = useHistory();
```
I don't know how to behave with hooks, we should check if every app is using newer version of React and if they are using it we should update the `react-router-dom` and `react` to newer version. But I'd rather use `withRouter` in here so we don't have any problems.

karelhala commented on packages/components/src/Components/NotAuthorized/NotAuthorized.js:
```diff
@@ -0,0 +1,42 @@
+};
+
+NotAuthorized.propTypes = {
+    serviceName: PropTypes.string
```
I'd use `PropTypes.node` in here.

karelhala commented on packages/components/src/Components/NotAuthorized/NotAuthorized.js:
```diff
@@ -0,0 +1,42 @@
+};
+
+NotAuthorized.defaultProps = {
+    serviceName: undefined
```
No need to add defaultProps to `undefined` it is by default `undefined` if you want to have a value in here you could use empty string or null for instance.

karelhala commented on packages/components/src/Components/NotAuthorized/notAuthorized.scss:
```diff
@@ -0,0 +1,19 @@
+
+    .ins-c-text__sorry { max-width: 70%; }
+
+    h1, svg, h2 { @include rem('margin-bottom', 30px); }
```
Either use `var(--pf-global--spacer--lg)` or `var(--pf-global--spacer--xl)` depends on what fits it better.

akofink commented:
Thanks for the review, @karelhala! I think I've addressed all your comments. How do I get around this error? I can't even see the page locally to make sure it matches the mocks.

`Error: Invariant failed: You should not use <withRouter(NotAuthorized) /> outside a <Router>`

karelhala commented on packages/components/src/Components/NotAuthorized/NotAuthorized.js:
```diff
@@ -0,0 +1,38 @@
+    Title,
+    Button
+} from '@patternfly/react-core';
+import { withRouter } from "react-router";
```
Single quotes should be used here.

karelhala commented:
@akofink I tried to run test locally and it is working fine, no error thrown. I'll try to include it in app as well and see the outcome.

karelhala commented:
@akofink this will be problem with different context since newer versions of react-router-dom started to use context it won't connect properly to it. The best way here will be by adding `history` prop that is passed from parent component. If no `history` is passed redirect users to root. This way you don't have to use `withRouter` or `useHistory` in this component, parent component can import history with `useHistory` and pass it down to this component, sounds much cleaner to be honest.

akofink commented:
I've updated this PR to match the DeniedState from the rbac-ui https://github.com/RedHatInsights/insights-rbac-ui/blob/master/src/presentational-components/states/DeniedState.js. It's a shame they didn't add it here first, since it's needed for all apps.

dLobatog commented on packages/components/src/Components/NotAuthorized/NotAuthorized.js:
```diff
@@ -0,0 +1,41 @@
+        </EmptyStateBody>
+        {
+          document.referrer ?
+            <Button variant="primary" onClick={ () => history.back() }>Return to previous page</Button> :
```
Just a nitpick, I think what @karelhala mentioned before was that it'd be good to have a `history` prop you could pass to this component to go back in history within the app without having to reload

akofink replied:
@dLobatog This is what you originally requested (using the browser's `history`), and that's what rbac-ui does. It seems to work pretty well - try to navigate to edit user roles in CI with your account, and you'll see this work as written here in the rbac-ui. https://github.com/RedHatInsights/insights-rbac-ui/blob/master/src/presentational-components/states/DeniedState.js#L32 No React required! What do you think?
